fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/RetryTask.java

```diff
@@ -0,0 +1,238 @@
+        globalLogs.addAll(Arrays.asList(
+                LogMessageKeys.MAX_ATTEMPTS, maxAttempts
+        ));
+        return globalLogs;
+    }
+
+    private void logCurrentAttempt(States<T> states, long delay) {
+        if (logger.isWarnEnabled()) {
+            final KeyValueLogMessage message = KeyValueLogMessage.build("Retrying Exception",
+                    LogMessageKeys.CURR_ATTEMPT, states.currAttempt,
+                    LogMessageKeys.DELAY, delay);
+            message.addKeysAndValues(states.localLogs);
+            message.addKeysAndValues(additionalLogMessageKeyValues);
+            logger.warn(message.toString(), states.possibleException.get());
+        }
+        states.localLogs.clear();
+    }
+
+    private Throwable addLogsToException(States<T> states, Throwable e) {
+        LoggableException loggableException = (e instanceof LoggableException) ?
+                                              (LoggableException)e : new LoggableException(e);
+        return loggableException
+                .addLogInfo(states.globalLogs.toArray())
+                // Include the local logs of the last retry as well.
+                .addLogInfo(states.localLogs.toArray());
+    }
+
+    public static class States<T> {
+        @Nonnull public final AtomicReference<T> possibleResult = new AtomicReference<>();
+        @Nonnull public final AtomicReference<Throwable> possibleException = new AtomicReference<>();
+        @Nonnull public final AtomicInteger currAttempt = new AtomicInteger(0);
+        @Nonnull public final AtomicLong currDelayRangeMillis;
+        // Keep the logs for the current attempt.
+        @Nonnull public final List<Object> localLogs = new ArrayList<>();
+        // Keep the logs for the entire run, including the additional ones provided by caller.
+        @Nonnull public final List<Object> globalLogs;
+
+        States(long initDelayMillis, List<Object> globalLogs) {
+            this.currDelayRangeMillis = new AtomicLong(initDelayMillis);
+            this.globalLogs = globalLogs;
+        }
+    }
+
+    public static class Builder<T> {
+        @Nonnull private final Supplier<CompletableFuture<? extends T>> retriableTask;
+        private final int maxAttempts;
+
+        @Nonnull private Function<RetryTask.States<T>, CompletableFuture<Boolean>> possiblyRetry =
+                states -> (states.possibleResult == null) ? AsyncUtil.READY_TRUE : AsyncUtil.READY_FALSE;
+        @Nonnull private Consumer<RetryTask.States<T>> handleIfDoRetry = states -> { };
+        private long initDelayMillis = 0;
+        private long minDelayMillis = 0;
+        private long maxDelayMillis = 0;
+        @Nonnull private Logger logger = DEFAULT_LOGGER;
+        @Nonnull private List<Object> additionalLogMessageKeyValues = Collections.emptyList();
+        @Nonnull private Executor executor = FDB.DEFAULT_EXECUTOR;
+        @Nullable private Function<Throwable,RuntimeException> exceptionMapper = null;
```

Comment:
There should probably be a non-`null` default exception mapper. I think `FDBExceptions::wrapException`?